src/NHibernate/Hql/Ast/ANTLR/Tree/SelectClause.cs

```diff
@@ -494,7 +494,7 @@ private void InitAliases(List<ISelectExpression> selectExpressions)
 				}
 
 				var node = (IASTNode) e;
-				if (processedElements.Add(fromElement))
+				if (Walker.IsShallowQuery && node.Type == SqlGenerator.DOT || processedElements.Add(fromElement))
```

Comment:
The `processedElements` set serves a de-duplication purpose of selected entities. But, alone, it fails accounting for associations selected in shallow queries which will be seen as having the same fromElement although being actually different entities being selected.

Shallow queries may still have duplicated entities selected and cannot be entirely ignored for the de-duplication, if we want to keep the test `HqlDuplicateEntitySelectionSubQuery` as currently is. (I fail to understand why this tests is valid, why should we support the queries in this test, but I did not get answers about this [here](https://github.com/nhibernate/nhibernate-core/pull/2551#pullrequestreview-1941571476). And anyway, it has been release in 5.4, so, it is legacy.)

The `DOT` check allows to recognize cases in shallow queries where the same fromElement may be used by different expressions selecting different things, and a such, needing to be not falsely de-duplicated.

(I thought about checking if the expressions `e` where "the same" (having same Path) instead when sharing a common fromElement, but that would imply building a dictionary of sets instead of `processedElements`. I do not consider it is worth it.)

Reply:
> I fail to understand why this tests is valid, why should we support the queries in this test

To be honest, I don't remember why I've added such test. Looking back at it, I agree that the test shouldn't be added.

Perhaps we could add a TODO for removing test `HqlDuplicateEntitySelectionSubQuery` in version `5.5` and remove the `node.Type == SqlGenerator.DOT` part from the condition.